paddle/fluid/imperative/utils/tensor_utils.cpp

```diff
@@ -0,0 +1,211 @@
+PyObject* TensorToNumpy(const framework::Tensor& tensor) {
+  // check the place of tensor
+  CheckTensor(tensor);
+
+  // get meta data (type, shape e.g..) from tensor
+  auto dtype = GetDType(tensor);
+  auto sizes = DDimToNumpyShape(tensor.dims());
+  // strides should take the sizeof type in count
+  auto strides = DDimToNumpyShape(framework::stride(tensor.dims()));
+  size_t type_size = framework::SizeOfType(tensor.type());
+  for (auto& stride : strides) {
+    stride *= type_size;
+  }
+
+  // create numpy array
+  PyObject* array = PyArray_New(
+      &PyArray_Type, tensor.dims().size(), sizes.data(), dtype, strides.data(),
+      tensor.Holder()->ptr(), 0, NPY_ARRAY_BEHAVED, nullptr);
+
+  if (!array) {
+    return nullptr;
+  }
+
+  // TODO(minqiyang): we should avoid to release the tensor here,
+  // since we passed it to numpy.
+  // // create a VarBase py_object and set it to Numpy PyArrayObj
+  // if (PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(array.get()),
+  // make_var_base(tensor)) == -1) {
+  // return nullptr;
+  // }
+
+  return array;
+}
+
+framework::Tensor TensorFromNumpy(PyObject* obj) {
+  // check numpy ndarray
+  if (!PyArray_Check(obj)) {
+    PADDLE_THROW("Only np.ndarray is supported, got %s here",
+                 Py_TYPE(obj)->tp_name);
+  }
+  auto array = reinterpret_cast<PyArrayObject*>(obj);
+
+  // get the meta data of numpy ndarray
+  int ndim = PyArray_NDIM(array);
+  std::vector<int64_t> dims = NumpyShapeToVector(ndim, PyArray_DIMS(array));
+  // NOTE(minqiyang): after tensor's requested size calculation, we will throw
+  // the strides info away, because tensor will calculate strides itself.
+  std::vector<int64_t> strides =
+      NumpyShapeToVector(ndim, PyArray_STRIDES(array));
+
+  // calculate the requested size of tensor
+  size_t requested_size = 0U;
+  if (dims.size() >= 1 && strides.size() >= 1) {
+    requested_size = static_cast<size_t>(dims[0] * strides[0]);
+  }
+
+  // check byte order
+  if (!PyArray_EquivByteorders(PyArray_DESCR(array)->byteorder, NPY_NATIVE)) {
+    PADDLE_ENFORCE(
+        PyArray_EquivByteorders(PyArray_DESCR(array)->byteorder, NPY_NATIVE),
+        "numpy array's byte orders should keep the same with the native byte "
+        "order");
+  }
+
+  // get the data ptr and construct the memory allocator attribute
+  void* data_ptr = PyArray_DATA(array);
+  framework::Tensor tensor;
+  tensor.clear();
+  tensor.Resize(framework::make_ddim(dims));
+  tensor.mutable_data(platform::CPUPlace(), GetTensorType(PyArray_TYPE(array)),
+                      memory::Allocator::Attr::kNumpyShared, 0U);
+  // NOTE(minqiyang): must be CPUAllocation here, we will down cast this
+  // allocation to CPUAllocation
+  memory::allocation::CPUAllocation* cpu_allocation =
+      reinterpret_cast<memory::allocation::CPUAllocation*>(
+          tensor.Holder().get());
+  cpu_allocation->share_data_with(data_ptr, requested_size);
+
+  Py_INCREF(obj);
```

Comment:
Are you sure the object will be released correctly? Because I could not find where invoke `Py_DECREF`

Reply:
Yes, I guess the ```Py_DECREF``` should be put into the destructor of imperative's ```Variable```, but ```Variable``` is still in process.

Reply:
I don't think so.

Since you carry the numpy memory in a tensor. It should be tensor's responsibility to correctly release the memory.

You need to write a `NumpyAllocation` to manage that.

Reply:
The `NumpyAllocation` does not need an allocator, since it is allocated by `Python`. But need to hold the `PyObject` and maintain its reference count.

Reply:
Oh, Perhaps you need a stub allocator to free `NumpyAllocation`.